Extra checking for IRQs being disabled

There is one server (Supermicro X7DWN) whcih occasionally falls over various
assertions to do with interrupts being disabled.

Given the symptoms, and that all other servers appear fine, it is likely to be
a hardware issue.

Put some more debugging to try and track down the point at which interrupts
get enabled.

diff --git a/xen/common/softirq.c b/xen/common/softirq.c
index 22e417a..9ee6e08 100644
--- a/xen/common/softirq.c
+++ b/xen/common/softirq.c
@@ -46,9 +46,21 @@ static void __do_softirq(unsigned long ignore_mask)
              || cpu_is_offline(cpu) )
             break;
 
+        if ( unlikely(!local_irq_is_enabled()) )
+        {
+            printk("CA-162193: IRQs disabled - pending mask %#lx\n", pending);
+            BUG();
+        }
+
         i = find_first_set_bit(pending);
         clear_bit(i, &softirq_pending(cpu));
         (*softirq_handlers[i])();
+
+        if ( unlikely(!local_irq_is_enabled()) )
+        {
+            printk("CA-162193: IRQs disabled after servicing softirq %u\n", i);
+            BUG();
+        }
     }
 }
 
